### PR TITLE
Add Logging Helper Functions

### PR DESCRIPTION
**Pull request by BenTaylor-TfN**
Describe Changes
---
Add module to make logging easier in CAF

Task Checklist
----
- [x] Have unittests been added (testing individual functions and their edge cases)
- [x] Have integration tests been added (if applicable, to test modules of functionality)
- [x] Updated RELEASE.md to reflect changes in PR
- [x] Have new dependencies been added?
  - [x] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [x] Have they been added to `setup.cfg`

### CODE REVIEW

**Comment by BenTaylor-TfN** [comment]
@wsp-mbuckley - I've moved some of the code over from NorMITs Demand - feel free to tailor to the use cases as you see fit.
I've made some minor changes to remove the demand specific parts, hopefully making it a little more generic.

I see you as the main user atm, so you probably have better vision of what will be useful here.

**Comment by BenTaylor-TfN** [comment]
Closes #22

**Comment by wsp-mbuckley** [comment]
@BenTaylor-TfN this is mostly ready for review, although tests aren't working with tox because of a ImportPathMismatchError, they work when I run pytest directly on my machine so I think it's something to do with the tox settings but I'm not sure

File: src/caf/toolkit/__init__.py [hunk 1]
```diff
@@ -3,5 +3,6 @@
 __version__ = _version.get_versions()["version"]
 
 # Alias
-from caf.toolkit import pandas_utils
+from caf.toolkit import pandas_utils, log_helpers
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/__init__.py:6]: What am I missing here?
Why are this aliases needed when these imports are already in this location?
I think I put the `pandas_utils` import here too and I can't remember why

File: tox.ini [hunk 2]
```diff
@@ -16,6 +16,8 @@ deps =
     -r {toxinidir}/requirements_dev.txt
 commands =
     pytest -n auto --basetemp={envtmpdir}
+    # Run serial tests on single process (these are ignored by default)
+    pytest -n0 --basetemp={envtmpdir} -m "serial"
```
Comment (BenTaylor-TfN) [inline tox.ini:20]: What's defined as a serial test here?

Comment (BenTaylor-TfN) [reply tox.ini:20]: Played around with the code and cracked this one now!

**Comment by BenTaylor-TfN** [comment]
> @BenTaylor-TfN this is mostly ready for review, although tests aren't working with tox because of a ImportPathMismatchError, they work when I run pytest directly on my machine so I think it's something to do with the tox settings but I'm not sure

I've been debating on removing tox for a while - it's supposed to make all the CI stuff easier but I keep hitting problems with it. Removing it would make one less point of failure.

File: pyproject.toml [hunk 3]
```diff
@@ -52,7 +52,11 @@ add-ignore = [
 ]
 
 [tool.pytest.ini_options]
-addopts = "--cov=caf.toolkit --cov-report=xml"
+addopts = '--cov=caf.toolkit --cov-report=xml --doctest-modules -m "not serial"'
+doctest_optionflags = "NORMALIZE_WHITESPACE"
+markers = [
+    "serial: marks tests to not run in parallel (will not run by default)",
+]
```
[the remark this answers is not in the document]
Comment (wsp-mbuckley) [reply pyproject.toml:59]: - `--doctest-modules`: turns on running doctests (examples in docstrings within caf.toolkit package)
- `-m "not serial"`: turns off any tests marked with serial
- `doctest_optionflags = "NORMALIZE_WHITESPACE"`: strips whitespace from results and answers when running doctests
- `markers`: registers custom 'serial' marker with pytest so it can be used to mark some (only 1 atm) tests to not run by default

File: src/caf/toolkit/__init__.py (hunk 1 above)
Comment (wsp-mbuckley) [reply src/caf/toolkit/__init__.py]: No idea I thought it was something to do with pytest or tox having issues but I've tested both and they work fine without so I've removed the import line in f7e37786a027d6d7e611ed64dac161737752a1ce

**Comment by wsp-mbuckley** [comment]
> > @BenTaylor-TfN this is mostly ready for review, although tests aren't working with tox because of a ImportPathMismatchError, they work when I run pytest directly on my machine so I think it's something to do with the tox settings but I'm not sure
> 
> I've been debating on removing tox for a while - it's supposed to make all the CI stuff easier but I keep hitting problems with it. Removing it would make one less point of failure.

I actually quite like tox, it helped me find an error with the `SystemInformation` class which could happen on a non-windows system also it has helped me catch some Python 3.9 errors which I didn't notice because I have 3.10 installed. I've found the errors with tox are caused by Pytests doctest and seem to be an issue in pytest when running in some sort of container (e.g. tox / docker) there's an environment variable which I think fixes the problem

File: pyproject.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (wsp-mbuckley) [reply pyproject.toml]: Moved the "not serial" flag to tox so its clearer than some tests are ran on a single process in tox, commit 2047fea88df05bfd05befc74bc73278e68b170c3

**Comment by wsp-mbuckley** [comment]
@BenTaylor-TfN addressed some of your comments and cleared up the pytest config (moved the serial / not serial stuff to tox where it makes more sense). Tox issues were caused by pytest doctest having problems running in containers and just disabled those warnings with an environment variable in the tox config

**Comment by wsp-mbuckley** [comment]
@BenTaylor-TfN, @isaac-tfn I've added some tests which check that the correct handlers are added to the warnings logger to make sure the function runs correctly and skipped the tests which actually check the warnings are in the log file because those still fail when running all tests (they pass when each test is ran individually).

I think that should be all the tests needed for this process so its ready for review / merging.

**Comment by BenTaylor-TfN** [comment]
I'll defer to @isaac-tfn for a review, alongside his implementation of this in Space

**Comment by isaac-tfn** [comment]
I'll pick this up when I have time - on leave Thursday Friday and have a few bits and pieces I've got to do, so it'll likely be next week but should be done by the time @wsp-mbuckley is back from Scotland

**Review by isaac-tfn — APPROVED** [review]
I haven't read every line of code but it all looks good, merging to master but I'll leave it to Ben to release.
